### ISSUE TITLE
Make it possible to chain together asserts  on a subject.

### ISSUE THREAD

**Issue by kbilsted** [fluentassertions/fluentassertions#1489]
### Description
Feature request: Make it possible to chain together asserts  on a subject.

This will have the following benefits

* more fluent typing. Typing “.A” can autocomplete “Also”, while typing “;” will not autocomplete “result.” Which is needed for the next test case
* Better group relevant assertions
* less technical, and supports stream of consciousness dapping down the asserts and reading them

### Complete minimal example reproducing the issue

Usually we assert we use the same variable `result` multiple times 

E.g.

```csharp
var result = a.Parts.GetStuff(parameters);
result.Code.Should().Be(1);
result.Explanation.Should().Be("");
result.Ids.Should().HaveCount(1);
```

I was wondering if this is easier to read and type

```csharp
var result = a.Parts.GetStuff(parameters);
result
    .With.Code.Should().Be(1)
    .Also.Explanation.Should().Be("")
    .Also.Ids.Should().HaveCount(1);
```

I guess the `With` is needed to bind what `Also` refers back to, similar to how `And` refers back to the last subject 

It also  allows better separation of logical grouping

```csharp
var result = a.Parts.GetStuff(parameters);
result
    .With.Code.Should().Be(1)
    .Also.Explanation.Should().Be("")
    .Also.Ids.Should().HaveCount(1);

result
    .With.Amount.Should().Be(400)
    .Also.AmountDetails.Should().Be("usd");
```

I prefer this syntax over

```cs
var result = a.Parts.GetStuff(parameters);
result.Should().BeEquivalientTo(new {
    Code = 1,
    Explanation = "",
    Ids = new []{ ?? }
});
```

since it has autocompletion and static typing

**Comment by jnyrup** [comment]
This is simply not possible.

```
var result = a.Parts.GetStuff(parameters);
result <-- we have an object of type Stuff at hand
    .With.Code.Should().Be(1) <--- we wrap it into NumericAssertions<int>
           ^-------- we're now have an int
    .Also.Explanation.Should().Be("")
            ^-- this would need to know that the int NumericAssertions<int> is wrapping, is a property on the Stuff type.
```

A way to write it completely fluent is with the `Match` assertion
```cs
result.Should()
    .Match<Stuff>(e => e.Code == 1)
    .And.Match<Stuff>(e => e.Explanation == "")
    .And.Match<Stuff>(e => e.Ids.Count == 1);
```

If you want to have the result of all three assertions, in case the first one fails, you can wrap them in an `AssertionScope`.
```cs
using(new AssertionScope())
{
	result
    	.With.Code.Should().Be(1)
	    .Also.Explanation.Should().Be("")
    	.Also.Ids.Should().HaveCount(1);
}
```

**Comment by kbilsted** [comment]
@jnyrup  thanks for a quick reply. I don't understand your explanation I'm afraid. The `With` is part of the fluent assertions, it sets on the threadlocalstorage a reference to the `Stuff` object, later `Also` looks up in the same storage and and continues the assert.

The alternatives you present are much harder for me to read and type - many more characters that are not autocompletable in VS. And require many characters that for non-us keyboards have terrible positions such as `< > = ;`  requiring extra presses to shift or alt. A "." is very fast and kicks in VS autocompletion.

**Comment by eNeRGy164** [comment]
I think @jnyrup tries to explain that your suggestion is technically not possible as the methods lose the scope you want to continue on.

**Comment by jnyrup** [comment]
My example was to show what type the type system sees, as you go a long the assertion you propose..
To put it in words.
* `result` is of type `Stuff`.
* `result.Code` is of type `int` (we no longer know if it came from `Stuff.Code`, `someArray.Length`, etc.)
* `result.Code.Should()` if of type `NumericAssertions<int>`
* `result.Code.Should().Be(1)` is of type `AndConstraint<NumericAssertions<int>>`
* `result.Code.Should().Be(1).And` is of type `NumericAssertions<int>`
* `result.Code.Should().Be(1).And.Explanation` is not compilable, as `Explanation` is not a property on * `NumericAssertions<int>`.

I can only agree that my alternatives are not as fluent, but at least they are technically possible.

I feel your struggle on programming on a Danish keyboard, I'm writing on one right now.

**Comment by jnyrup** [comment]
While I do like the idea of an even more fluent API I can't see how this is technically possible.
I'll close this issue for now, but feel free to re-open if you have new light to shed on this.
